File: zhao-picture-frontend/src/api/pictureController.ts
// @ts-ignore
/* eslint-disable */
import request from "@/request";

/** auditPicture POST /api/picture/audit/admin */
export async function auditPictureUsingPost(
  body: API.PictureAuditRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean_>("/api/picture/audit/admin", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** deletePictureById POST /api/picture/delete */
export async function deletePictureByIdUsingPost(
  body: API.DeleteRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseLong_>("/api/picture/delete", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** editPicture POST /api/picture/edit */
export async function editPictureUsingPost(
  body: API.PictureEditRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePictureVO_>("/api/picture/edit", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** getPictureVOById GET /api/picture/get/vo */
export async function getPictureVoByIdUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getPictureVOByIdUsingGETParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePictureVO_>("/api/picture/get/vo", {
    method: "GET",
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** createPictureOutPaintingTask POST /api/picture/out_painting/create_task */
export async function createPictureOutPaintingTaskUsingPost(
  body: API.CreatePictureOutPaintingTaskRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseCreateOutPaintingTaskResponse_>(
    "/api/picture/out_painting/create_task",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** getPictureOutPaintingTask GET /api/picture/out_painting/get_task */
export async function getPictureOutPaintingTaskUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getPictureOutPaintingTaskUsingGETParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseGetOutPaintingTaskResponse_>(
    "/api/picture/out_painting/get_task",
    {
      method: "GET",
      params: {
        ...params,
      },
      ...(options || {}),
    }
  );
}

/** select POST /api/picture/page/select/query */
export async function selectUsingPost(
  body: API.PictureQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePagePictureVO_>(
    "/api/picture/page/select/query",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** selectBySpace POST /api/picture/page/select/query/space */
export async function selectBySpaceUsingPost(
  body: API.PictureQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePagePictureVO_>(
    "/api/picture/page/select/query/space",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: body,
      ...(options || {}),
    }
  );
}

/** selectAdmin POST /api/picture/select/admin */
export async function selectAdminUsingPost(
  body: API.PictureQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePagePicture_>("/api/picture/select/admin", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** listPictureTagCategory GET /api/picture/tag_category */
export async function listPictureTagCategoryUsingGet(options?: {
  [key: string]: any;
}) {
  return request<API.BaseResponsePictureTagCategory_>(
    "/api/picture/tag_category",
    {
      method: "GET",
      ...(options || {}),
    }
  );
}

/** uploadPicture POST /api/picture/upload */
export async function uploadPictureUsingPost(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.uploadPictureUsingPOSTParams,
  body: {},
  file?: File,
  options?: { [key: string]: any }
) {
  const formData = new FormData();

  if (file) {
    formData.append("file", file);
  }

  Object.keys(body).forEach((ele) => {
    const item = (body as any)[ele];

    if (item !== undefined && item !== null) {
      if (typeof item === "object" && !(item instanceof File)) {
        if (item instanceof Array) {
          item.forEach((f) => formData.append(ele, f || ""));
        } else {
          formData.append(
            ele,
            new Blob([JSON.stringify(item)], { type: "application/json" })
          );
        }
      } else {
        formData.append(ele, item);
      }
    }
  });

  return request<API.BaseResponsePictureVO_>("/api/picture/upload", {
    method: "POST",
    params: {
      ...params,
    },
    data: formData,
    requestType: "form",
    ...(options || {}),
  });
}

/** uploadPictureByBatch POST /api/picture/upload/batch */
export async function uploadPictureByBatchUsingPost(
  body: API.PictureUploadByBatchRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseInt_>("/api/picture/upload/batch", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

/** uploadPictureByUrl POST /api/picture/upload/url */
export async function uploadPictureByUrlUsingPost(
  body: API.PictureUploadRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePictureVO_>("/api/picture/upload/url", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}

File: zhao-picture-backed/src/main/java/com/zhao/zhaopicturebacked/controller/PictureController.java
package com.zhao.zhaopicturebacked.controller;

import cn.hutool.core.lang.TypeReference;
import cn.hutool.core.util.StrUtil;
import cn.hutool.json.JSONUtil;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.zhao.zhaopicturebacked.annotation.AuthType;
import com.zhao.zhaopicturebacked.api.aliyun.AliYunApi;
import com.zhao.zhaopicturebacked.api.aliyun.CreateOutPaintingTaskResponse;
import com.zhao.zhaopicturebacked.api.aliyun.GetOutPaintingTaskResponse;
import com.zhao.zhaopicturebacked.cache.*;
import com.zhao.zhaopicturebacked.constant.BaseResponse;
import com.zhao.zhaopicturebacked.constant.UserConstant;
import com.zhao.zhaopicturebacked.domain.Picture;
import com.zhao.zhaopicturebacked.domain.Task;
import com.zhao.zhaopicturebacked.domain.User;
import com.zhao.zhaopicturebacked.enums.AuditStatusEnum;
import com.zhao.zhaopicturebacked.enums.CodeEnum;
import com.zhao.zhaopicturebacked.exception.BusinessException;
import com.zhao.zhaopicturebacked.model.PictureTagCategory;
import com.zhao.zhaopicturebacked.model.PictureVO;
import com.zhao.zhaopicturebacked.model.UserVO;
import com.zhao.zhaopicturebacked.request.DeleteRequest;
import com.zhao.zhaopicturebacked.request.picture.*;
import com.zhao.zhaopicturebacked.service.PictureService;
import com.zhao.zhaopicturebacked.service.TaskService;
import com.zhao.zhaopicturebacked.service.UserService;
import com.zhao.zhaopicturebacked.service.impl.PictureServiceImpl;
import com.zhao.zhaopicturebacked.task.aliyun.AiPictureRunner;
import com.zhao.zhaopicturebacked.utils.*;
import lombok.extern.slf4j.Slf4j;
import org.redisson.api.RLock;
import org.redisson.api.RedissonClient;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.util.DigestUtils;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import javax.annotation.Resource;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/picture")
@Slf4j
public class PictureController {
    @Resource
    private StringRedisTemplate stringRedisTemplate;

    @Resource
    private PictureService pictureService;

    @Resource
    private UserService userService;
    @Resource
    private RedisCacheStrategy redisCacheStrategy;
    @Resource
    private CaffeienCacheStrategy caffeienCacheStrategy;


    @Autowired
    private PictureServiceImpl pictureServiceImpl;


    @Resource
    private RedissonClient redissonClient;

    @Resource
    private AliYunApi aliYunApi;

    @Resource
    private TaskService taskService;


    /**
     * 创建 AI 扩图任务
     */
    @PostMapping("/out_painting/create_task")
    public BaseResponse<CreateOutPaintingTaskResponse> createPictureOutPaintingTask(
            @RequestBody CreatePictureOutPaintingTaskRequest createPictureOutPaintingTaskRequest,
            HttpServletRequest request) {
        if (createPictureOutPaintingTaskRequest == null || createPictureOutPaintingTaskRequest.getPictureId() == null) {
            throw new BusinessException(CodeEnum.PARAMES_ERROR);
        }
        UserVO loginUser = TokenUtil.getLoginUserVOFromCookie(request);
        CreateOutPaintingTaskResponse response = pictureService.createPictureOutPaintingTask(createPictureOutPaintingTaskRequest, loginUser);

        String taskId = response.getOutput().getTaskId();
        Task task = new Task();
        task.setTaskId(taskId);
        task.setTaskStatus(response.getOutput().getTaskStatus());
        task.setUserId(loginUser.getId());
        boolean b = taskService.saveOrUpdate(task);
        if (!b){
            ThrowUtil.throwBusinessException(CodeEnum.SYSTEM_ERROR,"保存任务失败");
        }
        Thread thread = new Thread(new AiPictureRunner(taskId,loginUser.getId()));
        thread.start();
        log.info("线程{}执行AI扩图异步任务",thread.getId());
        return ResultUtil.success(response);
    }

    /**
     * 查询 AI 扩图任务
     */
    @GetMapping("/out_painting/get_task")
    public BaseResponse<GetOutPaintingTaskResponse> getPictureOutPaintingTask(String taskId) {
        if(StrUtil.isBlank(taskId)){
            ThrowUtil.throwBusinessException(CodeEnum.PARAMES_ERROR,"taskId参数为空");
        }
        QueryWrapper<Task> taskQueryWrapper = new QueryWrapper<>();
        taskQueryWrapper.eq("task_id",taskId);
        Task task = taskService.getOne(taskQueryWrapper);
        GetOutPaintingTaskResponse getOutPaintingTaskResponse = new GetOutPaintingTaskResponse();
        GetOutPaintingTaskResponse.Output output = new GetOutPaintingTaskResponse.Output();
        BeanUtils.copyProperties(task,output);
        getOutPaintingTaskResponse.setOutput(output);
        if(task.getTaskStatus().equals("SUCCEEDED")){
            boolean b = taskService.removeById(task.getId());
            if(!b){
                log.error("删除任务失败");
            }
        }

        return ResultUtil.success(getOutPaintingTaskResponse);
    }




    /**
     * 通过文件上传图片
     * @param multipartFile
     * @param request
     * @return
     */
    @PostMapping("/upload")
    @AuthType(userType = UserConstant.USER)
    public BaseResponse<PictureVO> uploadPicture(@RequestPart("file") MultipartFile multipartFile, PictureUploadRequest pictureUploadRequest,HttpServletRequest request) {
        UserVO loginUserVO = TokenUtil.getLoginUserVOFromCookie(request);

        Long userId = loginUserVO.getId();

        PictureVO pictureVO = pictureService.uploadPicture(multipartFile,pictureUploadRequest, loginUserVO);


        User user = userService.getById(userId);
        UserVO userVO = UserUtil.getUserVOByUser(user);
        pictureVO.setUserVO(userVO);

       return ResultUtil.success(pictureVO);
    }

    /**
     * 通过url上传图片
     * @param
     * @param request
     * @return
     */
    @PostMapping("/upload/url")
    @AuthType(userType = UserConstant.USER)
    public BaseResponse<PictureVO> uploadPictureByUrl(@RequestBody PictureUploadRequest pictureUploadRequest,HttpServletRequest request) {
        UserVO loginUserVO = TokenUtil.getLoginUserVOFromCookie(request);

        Long userId = loginUserVO.getId();

        String fileUrl = pictureUploadRequest.getFileUrl();

        PictureVO pictureVO = pictureService.uploadPicture(fileUrl,pictureUploadRequest, loginUserVO);


        User user = userService.getById(userId);
        UserVO userVO = UserUtil.getUserVOByUser(user);
        pictureVO.setUserVO(userVO);
        return ResultUtil.success(pictureVO);
    }

    @PostMapping("/upload/batch")
    @AuthType(userType = UserConstant.ADMIN)
    public BaseResponse<Integer> uploadPictureByBatch(@RequestBody PictureUploadByBatchRequest pictureUploadByBatchRequest, HttpServletRequest request, HttpServletResponse response) {

        UserVO loginUserVO = TokenUtil.getLoginUserVOFromCookie(request);

        int count = pictureService.uploadPictureByBatch(pictureUploadByBatchRequest, loginUserVO);
        return ResultUtil.success(count,"上传成功");
    }



    /**
     * 删除图片
     * @param deleteRequest
     * @param request
     * @return
     */
    @PostMapping("/delete")
    @AuthType(userType = 0)
    public BaseResponse<Long> deletePictureById(@RequestBody DeleteRequest deleteRequest,HttpServletRequest request){
        UserVO loginUserVO = TokenUtil.getLoginUserVOFromCookie(request);
        Long delete = pictureService.deletePicture(deleteRequest,loginUserVO);
        return ResultUtil.success(delete,"删除成功");
    }


    /**
     * 分页查询图片（无登录可用）
     * @param pictureQueryRequest
     * @return
     */
    @PostMapping("/page/select/query")
    public BaseResponse<Page<PictureVO>> select(@RequestBody PictureQueryRequest pictureQueryRequest,HttpServletRequest request){

        pictureQueryRequest.setNullSpaceId( true);
        String servletPath = request.getServletPath();
        if(!servletPath.contains("admin")){
            pictureQueryRequest.setAuditStatus(AuditStatusEnum.REVIEW_PASS.getCode());
        }
        //根据PictureQueryRequest使用MD5加密构造缓存key
        String key = getCacheKey(pictureQueryRequest);
        Page<PictureVO> pictureVOPage = null;
        //查找Caffeine缓存
        String cache = caffeienCacheStrategy.getCache(key);
        if (cache != null){
            log.info("从Caffeine缓存里获取数据成功");
            pictureVOPage = JSONUtil.toBean(cache,new TypeReference<Page<PictureVO>>() {}, true);
            return ResultUtil.success(pictureVOPage);
        }
        //redis查找缓存
        cache = redisCacheStrategy.getCache(key);
        if(cache != null){
            log.info("从redis里获取数据成功");
            pictureVOPage = JSONUtil.toBean(cache, new TypeReference<Page<PictureVO>>() {}, true);
            //将缓存写入Caffeine缓存
            caffeienCacheStrategy.setCache(key,cache);
            return ResultUtil.success(pictureVOPage);
        }

        //锁可以再细一点，查询不同的分页用的锁也不同
        String jsonStr = JSONUtil.toJsonStr(pictureQueryRequest);
        String lockKeyMd5 = DigestUtils.md5DigestAsHex(jsonStr.getBytes(StandardCharsets.UTF_8));
        String lockkey = String.format("zhaopicture:controller:select_picture:%s:lock",lockKeyMd5);

        RLock lock = redissonClient.getLock(lockkey);
        try{
            //如果获取锁失败则返回
            if(!lock.tryLock(10,-1,TimeUnit.SECONDS)){
                log.info("获取锁失败");
                return ResultUtil.success(pictureVOPage);
            }

            //抢到了锁之后，再次查询缓存
            cache = caffeienCacheStrategy.getCache(key);
            if (cache != null){
                log.info("从Caffeine缓存里获取数据成功");
                pictureVOPage = JSONUtil.toBean(cache, new TypeReference<Page<PictureVO>>() {}, true);
                return ResultUtil.success(pictureVOPage);
            }
            cache = redisCacheStrategy.getCache(key);
            if(cache != null){
                log.info("从redis里获取数据成功");
                pictureVOPage = JSONUtil.toBean(cache, new TypeReference<Page<PictureVO>>() {}, true);
                //将缓存写入Caffeine缓存
                caffeienCacheStrategy.setCache(key,cache);
                return ResultUtil.success(pictureVOPage);
            }
            //如果还是没有缓存，就需要查数据库了
            Page<Picture> picturePage = pictureService.selectPage(pictureQueryRequest);
            //如果没查到数据，就直接返回空列表,同时也将空列表缓存进redis和Caffeine，防止用户恶意访问不存在的数据,使得数据库压力变大
            List<Picture> pictureList = picturePage.getRecords();
            if(pictureList.size()==0 ){
                log.info("缓存空值");
                caffeienCacheStrategy.setCache(key,JSONUtil.toJsonStr(new Page<PictureVO>()));
                redisCacheStrategy.setCache(key,JSONUtil.toJsonStr(new Page<PictureVO>()),60*15,TimeUnit.SECONDS);
                return ResultUtil.success(new Page<PictureVO>());
            }
            List<PictureVO> pictureVOList = pictureList.stream().map(picture ->pictureServiceImpl.getPictureVOByPicture(picture)).collect(Collectors.toList());
            //优化点，PictureVO里的UserVO字段需要回库查询数据，很多图片的userId可能都是一样的，一样的userId我们没必要查询多遍
            //先将UserId进行去重
            Set<Long> userIdSet = pictureList.stream().map(Picture::getUserId).collect(Collectors.toSet());
            Map<Long, User> userMap = userService.listByIds(userIdSet).stream().collect(Collectors.toMap(User::getId, user -> user));
            pictureVOList.forEach(pictureVO -> {
                if (userMap.containsKey(pictureVO.getUserId())){
                    pictureVO.setUserVO(UserUtil.getUserVOByUser(userMap.get(pictureVO.getUserId())));
                }
            });
            pictureVOPage = new Page<>(picturePage.getCurrent(), picturePage.getSize(), picturePage.getTotal());
            pictureVOPage.setRecords(pictureVOList);
            //给Caffeine缓存数据
//            cache = JSONUtil.toJsonStr(pictureVOPage);
//            log.info("缓存数据");
//            caffeienCacheStrategy.setCache(key, cache);
//            //给redis设置缓存,并设置缓存过期时间
//            redisCacheStrategy.setCache(key, cache, 60*60, TimeUnit.SECONDS);

        }catch (Exception e){
            log.error("分页查询图片数据失败",e);
            ThrowUtil.throwBusinessException(CodeEnum.SYSTEM_ERROR,"系统错误");
        }finally {
            //要确保是自己的锁才释放，获取锁失败的人不能释放
            if(lock.isHeldByCurrentThread()){
                log.info("线程:{},unlock",Thread.currentThread());
                lock.unlock();
            }

        }
        return ResultUtil.success(pictureVOPage,"查询成功");
    }


    /**
     * 分页查询私有空间的图片
     * @param pictureQueryRequest
     * @return
     */
    @PostMapping("/page/select/query/space")
    public BaseResponse<Page<PictureVO>> selectBySpace(@RequestBody PictureQueryRequest pictureQueryRequest,HttpServletRequest request){
        pictureQueryRequest.setNullSpaceId( false);
        String servletPath = request.getServletPath();
        if(!servletPath.contains("admin")){
            pictureQueryRequest.setAuditStatus(AuditStatusEnum.REVIEW_PASS.getCode());
        }

        //要查数据库了
        Page<Picture> picturePage = pictureService.selectPageBySpace(pictureQueryRequest,request);
        //如果没查到数据，就直接返回空列表,同时也将空列表缓存进redis和Caffeine，防止用户恶意访问不存在的数据,使得数据库压力变大
        List<Picture> pictureList = picturePage.getRecords();
        if(pictureList.size()==0 ){
            return ResultUtil.success(new Page<PictureVO>());
        }
        List<PictureVO> pictureVOList = pictureList.stream().map(picture ->pictureServiceImpl.getPictureVOByPicture(picture)).collect(Collectors.toList());
        //优化点，PictureVO里的UserVO字段需要回库查询数据，很多图片的userId可能都是一样的，一样的userId我们没必要查询多遍
        //先将UserId进行去重
        Set<Long> userIdSet = pictureList.stream().map(Picture::getUserId).collect(Collectors.toSet());
        Map<Long, User> userMap = userService.listByIds(userIdSet).stream().collect(Collectors.toMap(User::getId, user -> user));
        pictureVOList.forEach(pictureVO -> {
            if (userMap.containsKey(pictureVO.getUserId())){
                pictureVO.setUserVO(UserUtil.getUserVOByUser(userMap.get(pictureVO.getUserId())));
            }
        });
        Page<PictureVO> pictureVOPage = new Page<>(picturePage.getCurrent(), picturePage.getSize(), picturePage.getTotal());
        pictureVOPage.setRecords(pictureVOList);


        return ResultUtil.success(pictureVOPage,"查询成功");
    }






    //构造缓存Key
    private static String getCacheKey(PictureQueryRequest pictureQueryRequest) {
        //先构造缓存key
        String jsonStr = JSONUtil.toJsonStr(pictureQueryRequest);
        String ma5HexJsonStr = DigestUtils.md5DigestAsHex(jsonStr.getBytes());
        String Key = "picture:pagePictureVOCache"+ma5HexJsonStr;
        return Key;
    }


    /**
     * 分页查询图片（管理员可用）
     * @param pictureQueryRequest
     * @return
     */
    @PostMapping("/select/admin")
    @AuthType(userType = UserConstant.ADMIN)
    public BaseResponse<Page<Picture>> selectAdmin(@RequestBody PictureQueryRequest pictureQueryRequest){

        pictureQueryRequest.setNullSpaceId(true);
        Page<Picture> picturePage = pictureService.selectPage(pictureQueryRequest);

        return ResultUtil.success(picturePage,"查询成功");
    }


    /**
     * 编辑图片
     * @param pictureEditRequest
     * @param request
     * @return
     */
    @PostMapping("/edit")
    @AuthType(userType = UserConstant.USER)
    public BaseResponse<PictureVO> editPicture(@RequestBody PictureEditRequest pictureEditRequest,HttpServletRequest request){
        UserVO loginUserVO = TokenUtil.getLoginUserVOFromCookie(request);
        PictureVO pictureVO = pictureService.editPicture(pictureEditRequest, loginUserVO);
        return ResultUtil.success(pictureVO);
    }

    /**
     * 获取图片详情
     * @param id
     * @return
     */
    @GetMapping("/get/vo")
    @AuthType(userType = UserConstant.USER)
    public BaseResponse<PictureVO> getPictureVOById(Long id){
        PictureVO pictureVO = pictureService.getPictureVOById(id);

        return ResultUtil.success(pictureVO);
    }

    /**
     * 审核图片
     * @param pictureAuditRequest
     * @param request
     * @return
     */
    @PostMapping("/audit/admin")
    @AuthType(userType = UserConstant.ADMIN)
    public BaseResponse<Boolean> auditPicture(@RequestBody PictureAuditRequest pictureAuditRequest, HttpServletRequest request){
        UserVO loginUserVO = TokenUtil.getLoginUserVOFromCookie(request);
        pictureService.auditPicture(pictureAuditRequest,loginUserVO);
        return ResultUtil.success(true);
    }


    /**
     * 预制标签和分类
     * @return
     */
    @GetMapping("/tag_category")
    public BaseResponse<PictureTagCategory> listPictureTagCategory() {
        PictureTagCategory pictureTagCategory = new PictureTagCategory();
        List<String> tagList = Arrays.asList("热门", "搞笑", "生活", "高清", "艺术", "校园", "背景", "简历", "创意");
        List<String> categoryList = Arrays.asList("模板", "电商", "表情包", "素材", "海报");
        pictureTagCategory.setTagList(tagList);
        pictureTagCategory.setCategoryList(categoryList);
        return ResultUtil.success(pictureTagCategory);
    }





}
